### PR TITLE
Proper runtime support

### PR DESCRIPTION
**Pull request by matthid**
Part of https://github.com/fsprojects/Paket/pull/2256 but extra PR to make it easier to review.

### CODE REVIEW

**Comment by matthid** [comment]
OK this is crazy. If I got that correct some nuget packages can specify additional dependencies for a specific runtime in a file called `runtime.json`.
 This looks like:

```
  "runtimes": {
    "win": {
      "Microsoft.Win32.Primitives": {
        "runtime.win.Microsoft.Win32.Primitives": "4.3.0"
      },
```

This means Paket needs to restore `runtime.win.Microsoft.Win32.Primitives` (which is not known through the regular dependency tree) to support the "win" rid (or any rid importing "win").

How do we do that in paket? I don't want to change the current resolver (as it is complicated enough). For now I'd try to implement this as a second phase (after regular restore has happened). This means that paket will ignore any `runtime.json` in the second layer (but doing this is insanity IMHO anyway)...

@forki please advise if that sounds like going into the wrong direction...

**Comment by forki** [comment]
Runtime.json is a brain fart. I never understood what people were thinking
to implement that hack.

Am 17.04.2017 16:13 schrieb "Matthias Dittrich" <notifications@github.com>:

**Comment by matthid** [comment]
And now what?

**Comment by matthid** [comment]
Ok this is an interesting start. I think I'll do the integration in the following way:

- [ ] Add a new "runtimes" directive to the deps and lockfile

```
source https://nuget.org/api/v2
runtimes aot,win

nuget Newtonsoft.Json redirects: force
```

- [ ] After regular install/update there will be another install/update step for the runtime dependencies for the enabled runtimes

This way restore doesn't need to be changed at all and the new packages will be in the regular resolution (I guess I need a new flag in the lockfile to "detect" if a package is in the regular resolution or if its picked because of runtime).

Still not sure if the default should be installing all packages for all runtimes OR no runtime at all. Currently I'm favoring all runtimes...

**Comment by cloudRoutine** [comment]
I'm not sure that it should default to installing for all runtimes (there are a ton of them) and in the absence of a specified runtime, isn't the default behavior to produce a `FDD` instead of a `SCD` ?

File: src/Paket.Core/Paket.Core.fsproj [hunk 1]
```diff
@@ -63,6 +63,11 @@
         <FSharpTargetsPath>$(MSBuildProgramFiles32)\Microsoft SDKs\F#\4.0\Framework\v4.0\Microsoft.FSharp.Targets</FSharpTargetsPath>
       </PropertyGroup>
     </When>
+    <When Condition="$(TargetFSharpCoreVersion) &gt;= 4.4.1.0 AND $(TargetFSharpCoreVersion) &lt; 4.4.1.0 ">
```
Comment (cloudRoutine) [inline src/Paket.Core/Paket.Core.fsproj]: These kinds of conditions don't work with xbuild, it's ridiculous that they don't and I'd rather write them this way too, but they end up causing more headaches than they're worth.

Comment (baronfel) [inline src/Paket.Core/Paket.Core.fsproj]: Maybe once mono 5.0 drops (replacing xbuild with msbuild 15 by default) this can be considered, though

Comment (cloudRoutine) [inline src/Paket.Core/Paket.Core.fsproj]: The target F# core version is a lot less useful when you're actually relying on paket and the fsharp.core nupkg to determine the version you're using

**Comment by matthid** [comment]
> I'm not sure that it should default to installing for all runtimes (there are a ton of them) and in the absence of a specified runtime, isn't the default behavior to produce a FDD instead of a SCD ?

Yes, but I'm not sure how they handle native dependencies in that case. From observing I'd say they ignore runtime deps (as in runtime.json) but they still copy stuff in the regular resolution....

File: src/Paket.Core/Paket.Core.fsproj (hunk 1 above)
Comment (matthid) [inline src/Paket.Core/Paket.Core.fsproj]: the travis build worked ...

File: src/Paket.Core/RuntimeGraph.fs [hunk 2]
```diff
@@ -0,0 +1,185 @@
+﻿namespace Paket
+
+open Domain
```
Comment (cloudRoutine) [inline src/Paket.Core/RuntimeGraph.fs]: @matthid can you add some doccoms for the functions in this file? I already know i'll have forgotten what half of this does in a few days

File: src/Paket.Core/RuntimeGraph.fs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                        | _ -> failwithf "unknown stuff in runtime-dependency: %O" kv.Value
+                    packageName, depsSpec)
+                |> Map.ofSeq
+              yield
+                  { Rid = rid
+                    InheritedRids = imports
+                    RuntimeDependencies = dependencies }
+          | _ -> failwithf "unknwn stuff in runtime-description: %O" t.Value ]
+
+    let readRuntimeGraphJ (json:JObject) =
```
Comment (cloudRoutine) [inline src/Paket.Core/RuntimeGraph.fs]: @matthid do you think it's possible to generalize this json graph structure a bit so we can use it to create the `project.assets.json` file as well?

Comment (matthid) [inline src/Paket.Core/RuntimeGraph.fs]: can you link an example?

Comment (cloudRoutine) [inline src/Paket.Core/RuntimeGraph.fs]: [project.assets.json & *.deps.json](https://gist.github.com/cloudRoutine/cf30ae79aed821d9052c419b13b8cff9)

**Comment by forki** [comment]
Why do we need runtime in deps file? Shouldn't it just install the runtime we just use?

**Comment by matthid** [comment]
> Why do we need runtime in deps file? Shouldn't it just install the runtime we just use?

Good point. But do you really want to read the fsproj file for info and then use that for resolving packages? What about multiple or no fsproj file? csproj? What about "old-style" fsproj? I don't even want to go into that. At least not in the first step (this PR) tbh...

**Comment by forki** [comment]
No not fsproj. I meant the following: install should lock everything, but
restore only needs to download the stuff for the target it currently runs
on

Am 18.04.2017 9:57 vorm. schrieb "Matthias Dittrich" <
notifications@github.com>:

**Comment by matthid** [comment]
What is "the target it currently runs on"?

**Comment by forki** [comment]
The OS if the machine

**Comment by matthid** [comment]
And what if I run on windows and want to publish my project for unix?

**Comment by forki** [comment]
Hook in restore into dotnet publish?

Am 18.04.2017 10:05 vorm. schrieb "Matthias Dittrich" <
notifications@github.com>:

**Comment by matthid** [comment]
So its the same problem you do another resolver step on "restore" to resolve the runtime deps. I think in pakets philosophy we should have already done that at that phase (IMHO) ... But it's up to you

**Comment by matthid** [comment]
I'd prefer the following when we get called by dotnet restore with an unknown runtime: 
 - If in CI/noninteractive mode we fail when a runtime is not in the deps file and we are called
 - In interactive mode we ask the user if we should update the deps file and do an "install"

**Comment by forki** [comment]
No not resolver. Only restore

For resolving (during install/update) we would check everything.


Am 18.04.2017 10:12 vorm. schrieb "Matthias Dittrich" <
notifications@github.com>:

**Comment by matthid** [comment]
@forki So we always resolve and lock every runtime down (during install/update). I'm fine with that, lets see how that works.
The lockfile might grow quite a bit with this.

But isn't install/update currently downloading the packages as well? Why would we change the behavior for the runtime deps?

I agree that our hook can have this behavior to only download whats needed. 

It is really confusing, and I don't know if there is a good way how this fits into our world view...
On the other side I don't expect this PR to be perfect, just a start to experiment how things might work...

**Comment by forki** [comment]
It does not fit our world view. That's the whole point. 
It's aspnet team rushing a stupid concept into DNX without telling anyone. Now whole ecosystem is stuck with this shit and nuget team is busy building verified accounts for Microsoft and Oracle but nobody else.

File: src/Paket.Core/AssemblyInfo.fs [hunk 4]
```diff
@@ -6,16 +6,16 @@ open System.Reflection
 [<assembly: AssemblyProductAttribute("Paket")>]
 [<assembly: AssemblyCompanyAttribute("Paket team")>]
 [<assembly: AssemblyDescriptionAttribute("A dependency manager for .NET with support for NuGet packages and git repositories.")>]
-[<assembly: AssemblyVersionAttribute("4.4.0")>]
-[<assembly: AssemblyFileVersionAttribute("4.4.0")>]
-[<assembly: AssemblyInformationalVersionAttribute("4.4.0")>]
+[<assembly: AssemblyVersionAttribute("4.4.1")>]
+[<assembly: AssemblyFileVersionAttribute("4.4.1")>]
+[<assembly: AssemblyInformationalVersionAttribute("4.4.1")>]
```
Comment (matthid) [inline src/Paket.Core/AssemblyInfo.fs]: @forki one suggestion across all your projects. Can you commit this immediately when you release (at the same time when you change release notes)? This way this doesn't always pop-up and makes merge conflicts.
Another way would be to `gitignore` those files and having only a default version (like 0.0.0.0) checked in for Visual Studio users...

File: src/Paket.Core/RuntimeGraph.fs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        m1
+        |> Map.toSeq
+        |> Seq.append (m2 |> Map.toSeq)
+        |> Seq.groupBy (fst)
+        |> Seq.map (fun (k, g) ->
+            match g |> Seq.map snd |> Seq.toList with
+            | [ a; b ] -> k, f a b
+            | [ a ] -> k, a
+            | _ -> failwithf "This should never happen")
+        |> Map.ofSeq
```
Comment (matthid) [inline src/Paket.Core/RuntimeGraph.fs]: Is there really no build-in function for that?

Comment (0x53A) [inline src/Paket.Core/RuntimeGraph.fs]: https://github.com/fsharp/fslang-suggestions/issues/560

Comment (matthid) [inline src/Paket.Core/RuntimeGraph.fs]: @0x53A thanks :)

File: src/Paket.Core/RuntimeGraph.fs (hunk 3 above)
Comment (matthid) [inline src/Paket.Core/RuntimeGraph.fs]: No I don't we should generalize here. But of course that info helps in generating those files as well.

File: src/Paket.Core/RuntimeGraph.fs (hunk 2 above)
Comment (matthid) [inline src/Paket.Core/RuntimeGraph.fs]: Yes, I added some with https://github.com/fsprojects/Paket/pull/2255/commits/c71d62b088ac10eb07e9a1e2d22708582572c6b2

File: src/Paket.Core/Files/LockFile.fs [hunk 6]
```diff
@@ -359,7 +364,14 @@ module LockFileParser =
...
                     parts.[1]
-            parts.[0],InstallSettings.Parse(optionsString)
+            let isRuntimeDependency, optionsString =
+                if optionsString.EndsWith ", isRuntimeDependency: true" then
+                    true, optionsString.Substring(0, optionsString.Length - ", isRuntimeDependency: true".Length)
+                elif optionsString.EndsWith "isRuntimeDependency: true" then
+                    assert (optionsString = "isRuntimeDependency: true")
+                    true, ""
+                else false, optionsString
+            parts.[0],isRuntimeDependency,InstallSettings.Parse(optionsString)
```
Comment (matthid) [inline src/Paket.Core/Files/LockFile.fs]: Maybe the parser and generator code should be cleaned up a bit in the future...

**Pull request merged**